docs/cudf/source/developer_guide/ARCHITECTURE.md

```diff
@@ -0,0 +1,208 @@
+
+A `Column` is composed of the following:
+
+- A **data type**, specifying the type of each element.
+- A **data buffer** that may store the data for the column elements.
+  Some column types do not have a data buffer, instead storing data in the children columns.
+- A **mask buffer** whose bits represent the validity (null or not null) of each element.
+  Nullability is a core concept in the Arrow data model.
+  Columns whose elements are all valid may not have a mask buffer.
+  Mask buffers are padded to 64 bytes.
+- Its **children**, a tuple of columns used to represent complex types such as structs or lists.
+- A **size** indicating the number of elements in the column.
+- An integer **offset** use to represent the first element of column that is the "slice" of another column.
+  The size of the column then gives the extent of the slice rather than the size of the underlying buffer.
+  A column that is not a slice has an offset of 0.
+
+More information about these fields can be found in the documentation of the
+[Apache Arrow Columnar Format](https://arrow.apache.org/docs/format/Columnar.html),
+which is what the cuDF `Column` is based on.
+
+As one example, a `NumericalColumn` with 1000 `int32` elements and containing nulls is composed of:
+
+1. A data buffer of size 4000 bytes (sizeof(int32) * 1000)
+2. A mask buffer of size 128 bytes (1000/8 padded to a multiple of 64
+   bytes)
+3. No children columns
+
+As another example, a `StringColumn` backing the Series `['do', 'you', 'have', 'any', 'cheese?']` is composed of:
+
+1. No data buffer
+2. No mask buffer as there are no nulls in the Series
+3. Two children columns:
+
+   - A column of UTF-8 characters
+     `['d', 'o', 'y', 'o', 'u', 'h' ..., '?']`
+   - A column of "offsets" to the characters column (in this case,
+     `[0, 2, 5, 9, 12, 19]`)
+
+
+### Data types
+
+cuDF uses [dtypes](https://numpy.org/doc/stable/reference/arrays.dtypes.html) to represent different types of data.
+Since efficient GPU algorithms require preexisting knowledge of data layouts,
+cuDF does not support the arbitrary `object` dtype, but instead defines a few custom types for common use-cases:
+- `ListDtype`: Lists where each element in every list in a Column is of the same type
+- `StructDtype`: Dicts where a given key always maps to values of the same type
+- `CategoricalDtype`: Analogous to the pandas categorical dtype except that the categories are stored in device memory
+- `DecimalDtype`: Fixed-point numbers
+- `IntervalDtype`: Intervals
+
+Note that there is a many-to-one mapping between data types and `Column` classes.
+For instance, all numerical types (floats and ints of different widths) are all managed using `NumericalColumn`.
+
+
+### Buffer
+
+Although a `Column` represents an Arrow-compliant data structure, it does not directly handle memory mangaement.
+That job is delegated to the `Buffer` class, which represents a device memory allocation that it _may or may not_ own.
```

Comment:
A `Buffer` never "owns" any memory. It always keeps a reference to the owner of the memory in its `_owner` attribute. That `_owner` can be an RMM `DeviceBuffer` (for RAPIDS-managed memory), or something else, like an object exposing `__cuda_array_interface__`.

Reply:
Good point, I'll reword this to avoid that implication.